src/main/java/com/aws/greengrass/deployment/activator/DefaultActivator.java

```diff
@@ -159,6 +168,33 @@ void rollback(DeploymentDocument deploymentDocument, CompletableFuture<Deploymen
         }
     }
 
+    /**
+     * Finds all services which are dependers of given broken services, directly or indirectly
+     * This method performs a breadth-first search, starting from the broken services and traversing through
+     * service dependencies.
+     * @param rollbackServices the set of rollback services to track
+     * @param brokenServiceNames the set of broken service names
+     * @return a set of all services depending on the broken services, including themselves
+     */
+    private Set<GreengrassService> findServiceDependers(final Set<GreengrassService> rollbackServices,
+                                                        final Set<String> brokenServiceNames) {
+        LinkedHashSet<GreengrassService> dependerServices = rollbackServices.stream()
+                .filter(service -> brokenServiceNames.contains(service.getName()))
+                .collect(Collectors.toCollection(LinkedHashSet::new));
+
+        // Breadth-first search to find all dependent services, staring from broken services
+        for (GreengrassService currentService : dependerServices) {
+            for (GreengrassService depender : currentService.getHardDependers()) {
```

Comment:
why only hard? We need all depender don't we?

Make sure our test includes both hard and soft dependencies.

Reply:
Soft does wait. It just doesn't restart in response to the dependency restarting.

Reply:
Oh, I'll change that

Reply:
Updated